File: app/src/main/java/edu/uw/listviewdemo/MainActivity.java
package edu.uw.listviewdemo;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v4.util.LruCache;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ImageButton;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.ImageLoader;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.NetworkImageView;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";

    private ArrayAdapter<String> adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final EditText searchField = (EditText)findViewById(R.id.txt_search);
        final ImageButton searchButton = (ImageButton)findViewById(R.id.btn_download);

        searchButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //downloadDinosaurData();

                String searchTerm = searchField.getText().toString();
                downloadMediaData(searchTerm);
            }
        });

        NetworkImageView imageView = (NetworkImageView)findViewById(R.id.img_remote);
        imageView.setImageUrl("https://dinoxp.com/wp-content/uploads/2016/02/dinort.png",
                RequestSingleton.getInstance(this).getImageLoader());


        //model!
//        String[] data = new String[99];
//        for(int i=99; i>0; i--){
//            data[99-i] = i+ " bottles of beer on the wall";
//        }
        ArrayList<String> data = new ArrayList<String>(); //empty data; ArrayList so modifiable

        //controller
        adapter = new ArrayAdapter<String>(
                this, R.layout.list_item, R.id.txtItem, data);

        //supports ListView or GridView
        AdapterView listView = (AdapterView)findViewById(R.id.list_view);
        listView.setAdapter(adapter);


        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                String movie = (String) parent.getItemAtPosition(position);
                Log.d(TAG, "You clicked on " + movie);
            }
        });
    }

    //download dinosaur names from dinoipsum
    private void downloadDinosaurData() {

        String urlString = "http://dinoipsum.herokuapp.com/api/?format=text&words=20&paragraphs=1";

        Request request = new StringRequest(Request.Method.GET, urlString,
                new Response.Listener<String>() {
                    public void onResponse(String response) {
                        Log.v(TAG, response);

                        String[] dinos = response.split(" "); //turn string into an array

                        adapter.clear();
                        for(String dino : dinos) {
                            adapter.add(dino);
                        }
                    }
                }, new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        Log.e(TAG, error.toString());
                    }
                });

//        RequestQueue requestQueue = Volley.newRequestQueue(getApplicationContext());
//        requestQueue.add(request);

        RequestSingleton.getInstance(this).add(request);
    }

    //search for media from iTunes
    private void downloadMediaData(String searchTerm) {

        String urlString = "";
        try {
            urlString = "https://itunes.apple.com/search?term="+URLEncoder.encode(searchTerm, "UTF-8")+"&media=movie&limit=25";
        }catch(UnsupportedEncodingException uee){
            Log.e(TAG, uee.toString());
            return;
        }

        Request request = new JsonObjectRequest(Request.Method.GET, urlString, null,
                new Response.Listener<JSONObject>() {
                    public void onResponse(JSONObject response) {
                        ArrayList<String> mediaTitles = new ArrayList<String>();

                        try {
                            //parse the JSON results
                            JSONArray results = response.getJSONArray("results");
                            for(int i=0; i<results.length(); i++){
                                JSONObject track = results.getJSONObject(i);
                                String title = track.getString("trackName");
                                String artist = track.getString("artistName");
                                mediaTitles.add(title+" ("+artist+")");
                                Log.v(TAG, mediaTitles.get(mediaTitles.size()-1));
                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                        adapter.clear();
                        for(String media : mediaTitles) {
                            adapter.add(media);
                        }
                    }
                }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                Log.e(TAG, error.toString());
            }
        });

        RequestSingleton.getInstance(this).add(request);
    }

    protected static class RequestSingleton {
        //the single instance of this singleton
        private static RequestSingleton instance;

        private RequestQueue requestQueue = null; //the singleton's RequestQueue
        private ImageLoader imageLoader = null;

        //private constructor; cannot instantiate directly
        private RequestSingleton(Context ctx){
            //create the requestQueue
            this.requestQueue = Volley.newRequestQueue(ctx.getApplicationContext());

            //create the imageLoader
            imageLoader = new ImageLoader(requestQueue,
                    new ImageLoader.ImageCache() {  //define an anonymous Cache object
                        //the cache instance variable
                        private final LruCache<String, Bitmap> cache = new LruCache<String, Bitmap>(20);

                        //method for accessing the cache
                        @Override
                        public Bitmap getBitmap(String url) {
                            return cache.get(url);
                        }

                        //method for storing to the cache
                        @Override
                        public void putBitmap(String url, Bitmap bitmap) {
                            cache.put(url, bitmap);
                        }
                    });
        }

        //call this "factory" method to access the Singleton
        public static RequestSingleton getInstance(Context ctx) {
            //only create the singleton if it doesn't exist yet
            if(instance == null){
                instance = new RequestSingleton(ctx);
            }

            return instance; //return the singleton object
        }

        //get queue from singleton for direct action
        public RequestQueue getRequestQueue() {
            return this.requestQueue;
        }

        //convenience wrapper method
        public <T> void add(Request<T> req) {
            requestQueue.add(req);
        }

        public ImageLoader getImageLoader() {
            return this.imageLoader;
        }
    }

}

File: app/src/main/java/edu/uw/listviewdemo/MainActivity.kt
package edu.uw.listviewdemo

import android.content.Context
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.util.LruCache
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.widget.*
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.*
import org.json.JSONException
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.*

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"
    private lateinit var adapter: ArrayAdapter<String> //will be assigned later (but will not be null!)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val searchField = findViewById<EditText>(R.id.txt_search)
        val searchButton = findViewById<ImageButton>(R.id.btn_download)

        searchButton.setOnClickListener {
            downloadDinosaurData();

            //val searchTerm = searchField.text.toString()
            //downloadMediaData(searchTerm)
        }

        //model!
        val data = mutableListOf<String>()
        for(i in 99 downTo 1){
            data.add(i.toString() + " bottles of beer on the wall")
        }

        //controller
        adapter = ArrayAdapter(this, R.layout.list_item, R.id.txtItem, data)

        //supports ListView or GridView
        val listView = findViewById<AdapterView<ArrayAdapter<String>>>(R.id.list_view) //wow
        listView.setAdapter(adapter)


        listView.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
            val item = parent.getItemAtPosition(position) as String
            Toast.makeText(this, "You clicked on $item", Toast.LENGTH_SHORT).show()
        }


        //load an image off the internet
        val imageView = findViewById<NetworkImageView>(R.id.img_remote)
        imageView.setImageUrl("https://dinoxp.com/wp-content/uploads/2016/02/dinort.png",
                VolleyService.getInstance(this).imageLoader)
    }

    //download dinosaur names from dinoipsum
    private fun downloadDinosaurData() {

        val urlString = "http://dinoipsum.herokuapp.com/api/?format=text&words=20&paragraphs=1"

        val request = StringRequest(Request.Method.GET, urlString,
                Response.Listener { response ->
                    Log.v(TAG, response)

                    val dinos = response.split(" ") //turn string into an array

                    adapter.clear()
                    for (dino in dinos) {
                        adapter.add(dino)
                    }
                }, Response.ErrorListener { error -> Log.e(TAG, error.toString()) })

        //        RequestQueue requestQueue = Volley.newRequestQueue(getApplicationContext());
        //        requestQueue.add(request);

        VolleyService.getInstance(this).add(request)
    }

    //search for media from iTunes
    private fun downloadMediaData(searchTerm: String) {

        var urlString = ""
        try {
            urlString = "https://itunes.apple.com/search?term=" + URLEncoder.encode(searchTerm, "UTF-8") + "&media=movie&limit=25"
        } catch (uee: UnsupportedEncodingException) {
            Log.e(TAG, uee.toString())
            return
        }

        val request = JsonObjectRequest(Request.Method.GET, urlString, null,
                Response.Listener { response ->
                    val mediaTitles = ArrayList<String>()

                    try {
                        //parse the JSON results
                        val results = response.getJSONArray("results")
                        for (i in 0 until results.length()) {
                            val track = results.getJSONObject(i)
                            val title = track.getString("trackName")
                            val artist = track.getString("artistName")
                            mediaTitles.add("$title ($artist)")
                            Log.v(TAG, mediaTitles[mediaTitles.size - 1])
                        }
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }

                    adapter.clear()
                    for (media in mediaTitles) {
                        adapter.add(media)
                    }
                }, Response.ErrorListener { error -> Log.e(TAG, error.toString()) })

        VolleyService.getInstance(this).add(request)
    }

    //A class to manage the Volley requestQueue as a singleton
    private class VolleyService private constructor(ctx: Context) { //private constructor; cannot instantiate directly
        companion object {
            private var instance: VolleyService? = null //the single instance of this singleton

            //call this "factory" method to access the Singleton
            fun getInstance(ctx: Context): VolleyService {
                //only create the singleton if it doesn't exist yet
                if (instance == null) {
                    instance = VolleyService(ctx)
                }

                return instance as VolleyService //force casting
            }
        }

        //from Kotlin docs
        val requestQueue: RequestQueue by lazy {
            Volley.newRequestQueue(ctx.applicationContext) //return the context-based requestQueue
        }

        //from Kotlin docs
        val imageLoader: ImageLoader by lazy {
            ImageLoader(requestQueue,
                    object : ImageLoader.ImageCache {
                        private val cache = LruCache<String, Bitmap>(20)
                        override fun getBitmap(url: String): Bitmap? {
                            return cache.get(url)
                        }
                        override fun putBitmap(url: String, bitmap: Bitmap) {
                            cache.put(url, bitmap)
                        }
                    })
        }

        //convenience wrapper method
        fun <T> add(req: Request<T>) {
            requestQueue.add(req)
        }

    }
}
